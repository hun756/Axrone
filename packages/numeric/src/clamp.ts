type BrandedNumber<T extends string> = number & { readonly __brand: T };

type FiniteNumber = BrandedNumber<'finite'>;

type ClampedNumber<Min extends number, Max extends number> = BrandedNumber<`clamped-${Min}-${Max}`>;

type IsFiniteNumber<T> = T extends number ? (number extends T ? false : true) : false;

type ValidateBounds<Min extends number, Max extends number> = [
    IsFiniteNumber<Min>,
    IsFiniteNumber<Max>,
] extends [true, true]
    ? [Min, Max]
    : never;

type IsWithinBounds<Value, Min, Max> = Value extends number
    ? Min extends number
        ? Max extends number
            ? true
            : false
        : false
    : false;

type ComputeClampedValue<Value, Min, Max> = Value extends number
    ? Min extends number
        ? Max extends number
            ? Value
            : never
        : never
    : never;

export type ClampResult<Value extends number, Min extends number, Max extends number> =
    ValidateBounds<Min, Max> extends [infer NormMin, infer NormMax]
        ? NormMin extends number
            ? NormMax extends number
                ? IsWithinBounds<Value, NormMin, NormMax> extends true
                    ? Value
                    : ComputeClampedValue<Value, NormMin, NormMax>
                : never
            : never
        : never;

export class NumericRangeError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'NumericRangeError';
        (this as any).__proto__ = NumericRangeError.prototype;
    }
}

const isFiniteNumber = (value: unknown): value is FiniteNumber =>
    typeof value === 'number' && value === value && isFinite(value);

const fastClamp = (value: number, min: number, max: number): number =>
    value < min ? min : value > max ? max : value;

const validateNumber = (value: unknown, context: string): void => {
    if (process.env.NODE_ENV !== 'production' && !isFiniteNumber(value)) {
        throw new NumericRangeError(`${context} must be a finite number`);
    }
};

const validateBounds = (min: number, max: number): void => {
    if (process.env.NODE_ENV !== 'production') {
        validateNumber(min, 'Minimum bound');
        validateNumber(max, 'Maximum bound');
    }
};

export function clamp<const Min extends number, const Max extends number>(
    value: number,
    min: Min,
    max: Max
): ValidateBounds<Min, Max> extends [infer NormMin, infer NormMax]
    ? NormMin extends number
        ? NormMax extends number
            ? ClampedNumber<NormMin, NormMax>
            : never
        : never
    : never;

export function clamp(value: number, min: number, max: number): number;

export function clamp(value: number, min: number, max: number): number {
    validateNumber(value, 'Value');
    validateBounds(min, max);

    return min <= max ? fastClamp(value, min, max) : fastClamp(value, max, min);
}

export function createBoundedClamp<const Min extends number, const Max extends number>(
    min: Min,
    max: Max
): ValidateBounds<Min, Max> extends [infer NormMin, infer NormMax]
    ? NormMin extends number
        ? NormMax extends number
            ? (value: number) => ClampedNumber<NormMin, NormMax>
            : never
        : never
    : never;

export function createBoundedClamp(min: number, max: number): (value: number) => number;

export function createBoundedClamp(min: number, max: number) {
    validateBounds(min, max);
    const [lowerBound, upperBound] = min <= max ? [min, max] : [max, min];

    return (value: number): number => {
        validateNumber(value, 'Value');
        return fastClamp(value, lowerBound, upperBound);
    };
}

export const clampUnsafe = (value: number, min: number, max: number): number =>
    fastClamp(value, min <= max ? min : max, min <= max ? max : min);

export const clampInt = (value: number, min: number, max: number): number => {
    const intValue = value | 0;
    const intMin = min | 0;
    const intMax = max | 0;
    const [lower, upper] = intMin <= intMax ? [intMin, intMax] : [intMax, intMin];
    return fastClamp(intValue, lower, upper);
};

export const clamp01 = (value: number): number => fastClamp(value, 0, 1);

export const clamp0255 = (value: number): number => fastClamp(value, 0, 255);

export const clampNegOneOne = (value: number): number => fastClamp(value, -1, 1);

export type { FiniteNumber, ClampedNumber, BrandedNumber, ValidateBounds, IsFiniteNumber };
