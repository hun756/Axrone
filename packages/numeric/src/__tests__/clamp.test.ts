import { clamp, createBoundedClamp, NumericRangeError } from '../clamp';

describe('Numeric Clamp Utility', () => {
    describe('clamp()', () => {
        test.each([
            [5, 0, 10, 5],
            [0, 0, 10, 0],
            [10, 0, 10, 10],
            [-5, 0, 10, 0],
            [15, 0, 10, 10],
            [5, 10, 0, 5],
            [-10, -20, -5, -10],
            [0.5, 0.1, 0.9, 0.5],
            [Number.MAX_SAFE_INTEGER - 10, Number.MAX_SAFE_INTEGER - 100, Number.MAX_SAFE_INTEGER, Number.MAX_SAFE_INTEGER - 10], // Large numbers
        ])('clamp(%f, %f, %f) should return %f', (value, min, max, expected) => {
            expect(clamp(value, min, max)).toBe(expected);
        });

        test('automatically normalizes reversed min/max bounds', () => {
            expect(clamp(5, 10, 0)).toBe(5);
            expect(clamp(15, 10, 0)).toBe(10);
            expect(clamp(-5, 10, 0)).toBe(0);
        });

        test('handles special numeric edge cases', () => {
            expect(clamp(0, -10, 10)).toBe(0);

            const value = 0.1 + 0.2; // 0.30000000000000004 due to floating point
            expect(clamp(value, 0, 0.3)).toBeCloseTo(0.3);

            expect(clamp(42, 42, 42)).toBe(42);
        });

        test('throws NumericRangeError for non-finite inputs', () => {
            // Non-finite value
            expect(() => clamp(NaN, 0, 10)).toThrow(NumericRangeError);
            expect(() => clamp(Infinity, 0, 10)).toThrow(NumericRangeError);
            expect(() => clamp(-Infinity, 0, 10)).toThrow(NumericRangeError);

            // Non-finite bounds
            expect(() => clamp(5, NaN, 10)).toThrow(NumericRangeError);
            expect(() => clamp(5, 0, NaN)).toThrow(NumericRangeError);
            expect(() => clamp(5, Infinity, 10)).toThrow(NumericRangeError);
            expect(() => clamp(5, 0, -Infinity)).toThrow(NumericRangeError);
        });

        test('error messages are descriptive and helpful', () => {
            expect(() => clamp(NaN, 0, 10)).toThrow(/Value must be a finite number/);
            expect(() => clamp(5, NaN, 10)).toThrow(/Bounds must be finite numbers/);
        });
    });

    // Bounded clamp factory function
    describe('createBoundedClamp()', () => {
        test('creates a reusable clamping function with fixed bounds', () => {
            const boundTo0_100 = createBoundedClamp(0 as number, 100 as number);

            expect(typeof boundTo0_100).toBe('function');
            expect(boundTo0_100(50)).toBe(50);
            expect(boundTo0_100(-10)).toBe(0);
            expect(boundTo0_100(200)).toBe(100);
        });

        test('factory correctly normalizes reversed bounds', () => {
            const boundTo0_100 = createBoundedClamp(100 as number, 0 as number);

            expect(boundTo0_100(50)).toBe(50);
            expect(boundTo0_100(-10)).toBe(0);
            expect(boundTo0_100(200)).toBe(100);
        });

        test('throws during factory creation if bounds are invalid', () => {
            expect(() => createBoundedClamp(NaN, 100)).toThrow(NumericRangeError);
            expect(() => createBoundedClamp(0, Infinity)).toThrow(NumericRangeError);
        });

        test('throws when clamping invalid values with factory function', () => {
            const boundTo0_100 = createBoundedClamp(0 as number, 100 as number);
            expect(() => boundTo0_100(NaN)).toThrow(NumericRangeError);
            expect(() => boundTo0_100(Infinity)).toThrow(NumericRangeError);
        });

        test('handles repeated clamping operations consistently', () => {
            const boundTo0_100 = createBoundedClamp(0 as number, 100 as number);
            const values = [-10, 0, 50, 100, 200];
            const expected = [0, 0, 50, 100, 100];

            const results = values.map(boundTo0_100);
            expect(results).toEqual(expected);
        });
    });

    // Performance comparison (simple benchmark)
    describe('Performance', () => {
        test('bounded clamp is more efficient for repeated operations', () => {
            // Setup test data
            const testSize = 1000;
            const values = Array.from({ length: testSize }, (_, i) => i * 10 - 5000);

            // Standard clamp approach
            const standardStart = performance.now();
            const standardResults = values.map(v => clamp(v, 0, 1000));
            const standardEnd = performance.now();

            // Bounded clamp approach
            const boundedStart = performance.now();
            const boundedClamp = createBoundedClamp(0 as number, 1000 as number);
            const boundedResults = values.map(boundedClamp);
            const boundedEnd = performance.now();

            expect(boundedResults).toEqual(standardResults);

            // We aren't making a strict assertion here as performance can vary,
            console.log(`Standard clamp: ${standardEnd - standardStart}ms`);
            console.log(`Bounded clamp: ${boundedEnd - boundedStart}ms`);

            expect(true).toBe(true);
        });
    });

    describe('Type Safety', () => {
        test('maintains input number type', () => {
            const int: number = 5;
            const result: number = clamp(int, 0, 10);
            expect(result).toBe(5);

            expect(typeof clamp(5, 0, 10)).toBe('number');
        });
    });

    describe('Error Objects', () => {
        test('NumericRangeError has correct structure', () => {
            try {
                clamp(NaN, 0, 10);
                fail('Expected error was not thrown');
            } catch (error) {
                if (error instanceof NumericRangeError) {
                    expect(error).toBeInstanceOf(NumericRangeError);
                    expect(error.name).toBe('NumericRangeError');
                    expect(error.message).toBe('Value must be a finite number');
                    expect(error).toMatchSnapshot();
                } else {
                    fail('Error is not an instance of NumericRangeError');
                }
            }
        });
    });

    describe('Integration', () => {
        test('works with Array methods', () => {
            const values = [1, 5, 10, 15, 20];
            const boundedValues = values.map(v => clamp(v, 5, 15));
            expect(boundedValues).toEqual([5, 5, 10, 15, 15]);
        });
    });
});
