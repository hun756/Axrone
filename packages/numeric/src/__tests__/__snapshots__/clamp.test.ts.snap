// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Numeric Clamp Utility Error Objects NumericRangeError has correct structure 1`] = `[NumericRangeError: Value must be a finite number]`;
